File: banchus-common/src/main/java/pe/nanamochi/banchus/util/Security.java
package pe.nanamochi.banchus.util;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Base64;

public final class Security {

  private Security() {}

  private static final SecureRandom secureRandom = new SecureRandom();
  private static final Base64.Encoder urlEncoder = Base64.getUrlEncoder().withoutPadding();

  public static String getMd5(final String plaintext) {
    try {
      MessageDigest m = MessageDigest.getInstance("MD5");
      byte[] digest = m.digest(plaintext.getBytes());
      StringBuilder sb = new StringBuilder();
      for (byte b : digest) {
        sb.append(Integer.toHexString((b & 0xFF) | 0x100), 1, 3);
      }
      return sb.toString();
    } catch (NoSuchAlgorithmException e) {
      throw new RuntimeException("CRITICAL: MD5 algorithm not found in JVM", e);
    }
  }

  public static String getMd5FromBytes(byte[] data) {
    try {
      MessageDigest m = MessageDigest.getInstance("MD5");
      byte[] digest = m.digest(data);
      StringBuilder sb = new StringBuilder();
      for (byte b : digest) {
        sb.append(Integer.toHexString((b & 0xFF) | 0x100), 1, 3);
      }
      return sb.toString();
    } catch (NoSuchAlgorithmException e) {
      throw new RuntimeException(e);
    }
  }

  public static String generateToken(int byteLength) {
    byte[] bytes = new byte[byteLength];
    secureRandom.nextBytes(bytes);
    return urlEncoder.encodeToString(bytes);
  }
}

File: banchus-common/src/main/kotlin/pe/nanamochi/banchus/util/Security.kt
package pe.nanamochi.banchus.util

import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64

object Security {
    private val secureRandom = SecureRandom()
    private val urlEncoder = Base64.getUrlEncoder().withoutPadding()

    fun generateToken(byteLength: Int): String =
        ByteArray(byteLength).also(secureRandom::nextBytes).let(urlEncoder::encodeToString)

    internal fun calculateMd5(data: ByteArray): String {
        val digest = MessageDigest.getInstance("MD5").digest(data)
        return digest.joinToString("") { "%02x".format(it) }
    }
}

fun String.toMd5(): String = Security.calculateMd5(this.toByteArray())

fun ByteArray.toMd5(): String = Security.calculateMd5(this)
